### ISSUE TITLE
Slice of structs?

### ISSUE THREAD

**Issue by quantum1423** [lunixbochs/struc#30]
This fails:

````
type srvEdgeResp struct {
	Num    int `struc:"uint8,sizeof=Neighs"`
	Neighs []Neighbor
}

type Neighbor struct {
	Address net.IP              `struc:"[4]byte"`
	Port    int                 `struc:"uint16"`
	Secret  []byte              `struc:"[32]byte"`
	PubKey  natrium.EdDSAPublic `struc:"[32]byte"`
}
````

with the error `reflect: call of reflect.Value.NumField on slice Value`.

Am I doing something horribly wrong, or is this a limitation of the library? It'll be quite annoying if this cannot be done at all.

**Comment by lunixbochs** [comment]
This will probably be a trivial fix. I need to generally support embedded structs better (issues #1 and #4 as well). I'll take a look tonight!

**Comment by lunixbochs** [comment]
This is partially implemented in the slice-struct branch (pack works, unpack does not).

**Comment by quantum1423** [comment]
My [current project](https://github.com/ProjectNiwl/sagiri) is blocked on this thing working, so it would be nice if this feature be completed and merged as early as possible. I'm implementing a custom protocol which makes extensive use of repeating structures many times.

**Comment by lunixbochs** [comment]
I fixed packing, currently working to make unpacking work.

**Comment by lunixbochs** [comment]
I just pushed an implementation. Can you test it? I'd also like a performance report, as there are some places I can optimize if it's not fast enough for you.

**Comment by quantum1423** [comment]
It's fast enough for my particular use case, but maybe not in general? In my use case it's not even fast enough to saturate the network (30000 ns per operation), but it doesn't matter since that part of my system is not I/O bound (it's essentially an RPC protocol, where the server-side processing is very much CPU-bound).

But it might not be acceptable if it were some sort of bulk data transfer protocol.

**Comment by lunixbochs** [comment]
Yeah it should be closer to 2000ns/op, so I must not be caching the field parses. I'll leave this issue open as a note to optimize performance.

**Comment by lunixbochs** [comment]
Hmm, it is cached. How big are your struct slices? I added a benchmark against my nightmare case [example struct](https://github.com/lunixbochs/struc/blob/master/struc_test.go#L14) (which you can run with `go test -bench .` and got pretty reasonable numbers (with a slice of 8 nested structs):

    BenchmarkFullEncode-4     200000          6534 ns/op
    BenchmarkFullDecode-4     300000          4312 ns/op

**Comment by quantum1423** [comment]
It's the same thing I posted in the OP. I tested with `Neighs` having a length of 10.

**Comment by lunixbochs** [comment]
Can you run your benchmark again on the latest master?

**Comment by mpontillo** [comment]
I encountered this issue as well in [one of my projects](https://github.com/mpontillo/psedit/blob/2d09f0a9c576cc9c08736b9da7f7517ebac761ac/savefile.go); I had to work around this for now by reading data with `encoding/binary` (writing works). I tried to hack `struc` a bit, but as this is my first Go project, I didn't get very far.

I'm happy to help test a fix (and possibly write test cases when I have time).
